File: lib/widgets/alarm_list_view.dart
import 'package:flutter/material.dart';
import '../models/alarm_model.dart';
import '../utils/alarm_utils.dart';
import '../widgets/alarm_tile.dart';
import 'alarm_header.dart';

class AlarmListView extends StatelessWidget {
  final List<AlarmModel> alarms;
  final DateTime now;
  final void Function(int) onDelete;
  final void Function(int, AlarmModel) onUpdate;
  final void Function(AlarmModel) onAdd;
  final void Function(int, bool) onToggle;
  final void Function(BuildContext, AlarmModel, int) onTap;

  const AlarmListView({
    super.key,
    required this.alarms,
    required this.now,
    required this.onDelete,
    required this.onUpdate,
    required this.onAdd,
    required this.onToggle,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final enabledAlarms = alarms.where((a) => a.enabled).toList();
    enabledAlarms.sort((a, b) => _alarmDateTime(a.time).compareTo(_alarmDateTime(b.time)));

    final nextAlarm = enabledAlarms.isNotEmpty ? enabledAlarms.first : null;
    final nextAlarmMessage = nextAlarm != null
        ? getTimeUntilAlarm(nextAlarm.time, now)
        : '설정된 다음 알람이 없습니다';

    return Column(
      children: [
        AlarmHeader(
          now: now,
          nextAlarmMessage: nextAlarmMessage,
        ),
        Expanded(
          child: ListView.builder(
            padding: const EdgeInsets.symmetric(horizontal: 16),
            itemCount: alarms.length,
            itemBuilder: (context, index) {
              final alarm = alarms[index];
              return GestureDetector(
                onTap: () => onTap(context, alarm, index),
                child: AlarmTile(
                  time: formatTimeOfDay(alarm.time),
                  days: alarm.days.map(dayOfWeekToKor).toList(),
                  enabled: alarm.enabled,
                  onToggle: (val) => onToggle(index, val),
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  DateTime _alarmDateTime(TimeOfDay time) {
    final scheduled = DateTime(now.year, now.month, now.day, time.hour, time.minute);
    return scheduled.isBefore(now) ? scheduled.add(const Duration(days: 1)) : scheduled;
  }
}

File: app/src/main/java/lab/p4c/nextup/ui/widget/AlarmListView.kt
package lab.p4c.nextup.ui.widget

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import lab.p4c.nextup.domain.model.Alarm
import lab.p4c.nextup.util.*
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

@Composable
fun AlarmListView(
    alarms: List<Alarm>,
    now: Instant,
    onDelete: (Int) -> Unit,
    onUpdate: (Alarm) -> Unit,
    onAdd: (Alarm) -> Unit,
    onToggle: (Alarm, Boolean) -> Unit,
    onTap: (Alarm, Int) -> Unit,
    modifier: Modifier = Modifier
) {
    val nowZdt: ZonedDateTime = now.atZone(ZoneId.systemDefault())

    // Flutter와 동일: 활성 알람만 추려 다음 울릴 순서로 정렬
    val enabledAlarms = alarms.filter { it.enabled }.sortedBy {
        val today = nowZdt.withHour(it.hour).withMinute(it.minute).withSecond(0).withNano(0)
        val dt = if (today.isBefore(nowZdt)) today.plusDays(1) else today
        dt.toInstant().toEpochMilli()
    }

    val nextAlarm = enabledAlarms.firstOrNull()
    val nextAlarmMessage = if (nextAlarm != null)
        getTimeUntilAlarm(nextAlarm.hour, nextAlarm.minute, nowZdt)
    else
        "설정된 다음 알람이 없습니다"

    Column(modifier = modifier.fillMaxSize()) {
        AlarmHeader(now = now, nextAlarmMessage = nextAlarmMessage)

        LazyColumn(
            contentPadding = PaddingValues(horizontal = 16.dp, vertical = 8.dp),
            modifier = Modifier.fillMaxSize()
        ) {
            itemsIndexed(alarms, key = { _, a -> a.id }) { index, alarm ->
                val timeText = formatTimeOfDay(alarm.hour, alarm.minute)
                val days = alarm.days.map { dayOfWeekToKor(it) }

                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { onTap(alarm, index) }
                ) {
                    AlarmTile(
                        time = timeText,
                        days = days,
                        enabled = alarm.enabled,
                        onToggle = { checked -> onToggle(alarm, checked) }
                    )
                }
            }

        }
    }
}
